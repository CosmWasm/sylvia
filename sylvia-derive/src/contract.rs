use communication::api::Api;
use communication::enum_msg::EnumMessage;
use communication::executor::Executor;
use communication::instantiate_builder::InstantiateBuilder;
use communication::querier::Querier;
use communication::reply::Reply;
use communication::struct_msg::StructMessage;
use communication::wrapper_msg::GlueMessage;
use mt::MtHelpers;
use proc_macro2::TokenStream;
use quote::quote;
use syn::{GenericParam, ItemImpl};

use crate::parser::attributes::features::SylviaFeatures;
use crate::parser::attributes::msg::MsgType;
use crate::parser::variant_descs::AsVariantDescs;
use crate::parser::{
    assert_new_method_defined, ContractErrorAttr, Custom, OverrideEntryPoint,
    ParsedSylviaAttributes,
};
use crate::types::interfaces::Interfaces;
use crate::types::msg_variant::MsgVariants;

mod communication;
mod mt;

/// Preprocessed `contract` macro input for struct impl block.
///
/// Generates:
///     - [Messages](https://cosmwasm-docs.vercel.app/sylvia/macros/generated-types/message-types#contract-messages)
///         - InstantiateMsg
///         - ExecMsg
///         - QueryMsg
///         - SudoMsg
///         - MigrateMsg
///         - ContractExecMsg
///         - ContractQueryMsg
///         - ContractSudoMsg
///     - [MultiTest](https://cosmwasm-docs.vercel.app/sylvia/macros/generated-types/multitest) helpers
///     - [Querier](https://cosmwasm-docs.vercel.app/cw-multi-test) trait implementation
///     - [Executor](https://cosmwasm-docs.vercel.app/cw-multi-test) trait implementation
///     - Api trait implementation
pub struct ContractInput<'a> {
    item: &'a ItemImpl,
    generics: Vec<&'a GenericParam>,
    error: ContractErrorAttr,
    custom: Custom,
    override_entry_points: Vec<OverrideEntryPoint>,
    interfaces: Interfaces,
    _sv_features: SylviaFeatures,
}

impl<'a> ContractInput<'a> {
    pub fn new(item: &'a ItemImpl) -> Self {
        assert_new_method_defined(item);

        let generics = item.generics.params.iter().collect();
        let parsed_attrs = ParsedSylviaAttributes::new(item.attrs.iter());
        let error = parsed_attrs.error_attrs.unwrap_or_default();
        let custom = parsed_attrs.custom_attr.unwrap_or_default();
        let sv_features = parsed_attrs.sv_features;
        let override_entry_points = parsed_attrs.override_entry_point_attrs;
        let interfaces = Interfaces::new(item);

        Self {
            item,
            generics,
            error,
            custom,
            override_entry_points,
            interfaces,
            _sv_features: sv_features,
        }
    }

    /// Processes the input and generates the contract code.
    pub fn process(&self) -> TokenStream {
        let Self {
            item,
            generics,
            custom,
            ..
        } = self;
        let multitest_helpers = self.emit_multitest_helpers();
        let messages = self.emit_messages();
        let contract_api = Api::new(item, generics, custom).emit();
        let querier = self.emit_querier();
        let executor = self.emit_executor();
        let reply = self.emit_reply();
        let instantiate_builder = self.emit_instantiate_builder_trait();

        quote! {
            pub mod sv {
                use super::*;

                #messages

                #multitest_helpers

                #querier

                #executor

                #reply

                #contract_api

                #instantiate_builder
            }
        }
    }

    fn emit_messages(&self) -> TokenStream {
        let instantiate = self.emit_struct_msg(MsgType::Instantiate);
        let migrate = self.emit_struct_msg(MsgType::Migrate);
        let exec_impl = self.emit_enum_msg(MsgType::Exec);
        let query_impl = self.emit_enum_msg(MsgType::Query);
        let sudo_impl = self.emit_enum_msg(MsgType::Sudo);
        let exec = self.emit_glue_msg(MsgType::Exec);
        let query = self.emit_glue_msg(MsgType::Query);
        let sudo = self.emit_glue_msg(MsgType::Sudo);

        quote! {
            #instantiate

            #exec_impl

            #query_impl

            #sudo_impl

            #migrate

            #exec

            #query

            #sudo
        }
    }

    fn emit_struct_msg(&self, msg_ty: MsgType) -> TokenStream {
        StructMessage::new(self.item, msg_ty, &self.generics, &self.error, &self.custom)
            .map_or(quote! {}, |msg| msg.emit())
    }

    fn emit_enum_msg(&self, msg_ty: MsgType) -> TokenStream {
        EnumMessage::new(self.item, msg_ty, &self.generics, &self.error, &self.custom).emit()
    }

    fn emit_glue_msg(&self, msg_ty: MsgType) -> TokenStream {
        GlueMessage::new(
            self.item,
            msg_ty,
            &self.error,
            &self.custom,
            &self.interfaces,
        )
        .emit()
    }

    fn emit_multitest_helpers(&self) -> TokenStream {
        if !cfg!(feature = "mt") {
            return quote! {};
        }

        let Self {
            item,
            custom,
            override_entry_points,
            ..
        } = self;

        let generic_params = &self.generics;
        MtHelpers::new(item, generic_params, custom, override_entry_points.clone()).emit()
    }

    fn emit_executor(&self) -> TokenStream {
        let item = self.item;
        let variants = MsgVariants::new(item.as_variants(), MsgType::Exec, &[], &None);

        Executor::new(item.generics.clone(), *item.self_ty.clone(), variants).emit()
    }
    fn emit_querier(&self) -> TokenStream {
        let item = self.item;
        let variants = MsgVariants::new(item.as_variants(), MsgType::Query, &[], &None);

        Querier::new(item.generics.clone(), *item.self_ty.clone(), variants).emit()
    }

    fn emit_reply(&self) -> TokenStream {
        let variants = MsgVariants::new(self.item.as_variants(), MsgType::Reply, &[], &None);

        Reply::new(self.item, &self.generics, &variants).emit()
    }

    fn emit_instantiate_builder_trait(&self) -> TokenStream {
        let variants = MsgVariants::new(
            self.item.as_variants(),
            MsgType::Instantiate,
            &self.generics,
            &self.item.generics.where_clause,
        );
        let where_clause = variants.where_clause();

        match variants.get_only_variant() {
            Some(variant) => InstantiateBuilder::new(
                *self.item.self_ty.clone(),
                variants.used_generics(),
                &where_clause,
                variant,
            )
            .emit(),
            None => quote! {},
        }
    }
}
