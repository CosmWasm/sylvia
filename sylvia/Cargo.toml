[package]
name = "sylvia"
version = { workspace = true }
authors = ["Bart≈Çomiej Kuras <bart.k@confio.gmbh>"]
edition = "2021"
description = "Framework for generating trait based CosmWasm smart contracts"
license = "Apache-2.0"
repository = "https://github.com/CosmWasm/sylvia"
homepage = "https://cosmwasm.com"
keywords = ["cosmwasm", "blockchain", "confio"]
categories = ["api-bindings", "wasm"]
readme = "../README.md"

[features]
mt = ["sylvia-derive/mt", "dep:cw-multi-test", "dep:anyhow"]
stargate = ["cosmwasm-std/stargate", "cw-multi-test/stargate"]
iterator = ["cosmwasm-std/iterator"]
cosmwasm_1_1 = ["cosmwasm-std/cosmwasm_1_1", "cw-multi-test/cosmwasm_1_1"]
cosmwasm_1_2 = [
    "cosmwasm-std/cosmwasm_1_2",
    "cw-multi-test/cosmwasm_1_2",
    "sylvia-derive/cosmwasm_1_2",
    "cosmwasm_1_1",
]
cosmwasm_1_3 = [
    "cosmwasm-std/cosmwasm_1_3",
    "cw-multi-test/cosmwasm_1_3",
    "cosmwasm_1_2",
]
cosmwasm_1_4 = [
    "cosmwasm-std/cosmwasm_1_4",
    "cw-multi-test/cosmwasm_1_4",
    "cosmwasm_1_3",
]
cosmwasm_2_0 = [
    "cosmwasm-std/cosmwasm_2_0",
    "cw-multi-test/cosmwasm_2_0",
    "cosmwasm_1_4",
]
cosmwasm_2_1 = [
    "cosmwasm-std/cosmwasm_2_1",
    "cw-multi-test/cosmwasm_2_1",
    "cosmwasm_2_0",
]
cosmwasm_2_2 = ["cosmwasm-std/cosmwasm_2_2", "cosmwasm_2_1"]

[dependencies]
sylvia-derive = { workspace = true }
cosmwasm-std = { workspace = true, features = ["staking"] }
cosmwasm-schema = { workspace = true }
schemars = { workspace = true }
serde = { workspace = true }
serde-cw-value = "0.7.0"
serde-json-wasm = "1.0.1"
konst = "0.3.15"
cw-multi-test = { workspace = true, features = ["staking"], optional = true }
anyhow = { workspace = true, optional = true }
cw-utils = { workspace = true }

[dev-dependencies]
cw-multi-test = { workspace = true, features = ["staking"] }
anyhow = { workspace = true }
cw-storage-plus = { workspace = true }
cw-utils = { workspace = true }
thiserror = { workspace = true }
trybuild = "1.0.101"
itertools = "0.13.0"

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--generate-link-to-definition"]
