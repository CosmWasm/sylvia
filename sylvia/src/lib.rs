#![cfg_attr(docsrs, feature(doc_cfg))]
//! Framework for creating CosmWasm Smart Contract with high-level abstraction layer.
//!
//! Most of implementation lies in `sylvia-derive` crate which is reexported here.

pub mod builder;
pub mod ctx;
pub mod into_response;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub mod multitest;
pub mod types;
pub mod utils;

#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use anyhow;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use cw_multi_test;
pub use cw_utils;
pub use sylvia_derive::{contract, entry_points, interface};
pub use {
    cosmwasm_schema as cw_schema, cosmwasm_std as cw_std, schemars, serde,
    serde_cw_value as serde_value,
};
