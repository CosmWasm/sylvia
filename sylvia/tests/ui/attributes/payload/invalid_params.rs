#![allow(unused_imports)]

use sylvia::contract;
use sylvia::ctx::{InstantiateCtx, ReplyCtx};
use sylvia::cw_std::{Binary, Reply, Response, StdResult};

pub struct Contract;

#[contract]
impl Contract {
    pub fn new() -> Self {
        Self
    }

    #[sv::msg(instantiate)]
    pub fn instantiate(&self, _ctx: InstantiateCtx) -> StdResult<Response> {
        Ok(Response::new())
    }

    #[sv::msg(reply, reply_on=success)]
    fn reply(
        &self,
        _ctx: ReplyCtx,
        #[sv::data(raw, opt)] _data: Option<Binary>,
        #[sv::payload(invalid)] _param: Option<Binary>,
    ) -> StdResult<Response> {
        Ok(Response::new())
    }
}

fn main() {}
