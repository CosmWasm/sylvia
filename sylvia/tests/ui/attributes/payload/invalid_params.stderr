error: Invalid payload parameter.
       = note: Expected [`raw`].
  --> tests/ui/attributes/payload/invalid_params.rs:25:23
   |
25 |         #[sv::payload(invalid)] _param: Option<Binary>,
   |                       ^^^^^^^
